packages/cli/src/services/role.service.ts

```diff
@@ -1,30 +1,124 @@
 } from '@n8n/db';
 import { Service } from '@n8n/di';
-import type { AllRoleTypes, Scope } from '@n8n/permissions';
-import { ALL_ROLES, combineScopes, getAuthPrincipalScopes, getRoleScopes } from '@n8n/permissions';
+import type { Scope } from '@n8n/permissions';
+import {
+	combineScopes,
+	getAuthPrincipalScopes,
+	getRoleScopes,
+	isBuiltInRole,
+} from '@n8n/permissions';
 import { UnexpectedError } from 'n8n-workflow';
 
 import { License } from '@/license';
+import { CreateRoleDto, RoleDTO, UpdateRoleDto } from '@n8n/api-types';
+import { NotFoundError } from '@/errors/response-errors/not-found.error';
+import { BadRequestError } from '@/errors/response-errors/bad-request.error';
 
 @Service()
 export class RoleService {
-	constructor(private readonly license: License) {}
+	constructor(
+		private readonly license: License,
+		private readonly roleRepository: RoleRepository,
+		private readonly scopeRepository: ScopeRepository,
+	) {}
 
-	getAllRoles() {
-		Object.values(ALL_ROLES).forEach((entries) => {
-			entries.forEach((entry) => {
-				entry.licensed = this.isRoleLicensed(entry.role);
-			});
+	private dbRoleToRoleDTO(role: Role): RoleDTO {
+		return {
+			...role,
+			scopes: role.scopes.map((s) => s.slug),
+			licensed: this.isRoleLicensed(role.slug),
+		};
+	}
+
+	async getAllRoles(): Promise<RoleDTO[]> {
+		const roles = await this.roleRepository.findAll();
+		return roles.map((r) => this.dbRoleToRoleDTO(r));
+	}
+
+	async getRole(slug: string): Promise<RoleDTO> {
+		const role = await this.roleRepository.findBySlug(slug);
+		if (role) {
+			return this.dbRoleToRoleDTO(role);
+		}
+		throw new NotFoundError('Role not found');
+	}
+
+	async removeCustomRole(slug: string) {
+		const role = await this.roleRepository.findBySlug(slug);
+		if (!role) {
+			throw new NotFoundError('Role not found');
+		}
+		if (role.systemRole) {
+			throw new BadRequestError('Cannot delete system roles');
+		}
+		await this.roleRepository.removeBySlug(slug);
+		return this.dbRoleToRoleDTO(role);
+	}
+
+	private async resolveScopes(scopeSlugs: string[] | undefined): Promise<DBScope[] | undefined> {
+		if (!scopeSlugs) {
+			return undefined;
+		}
+
+		if (scopeSlugs.length === 0) {
+			return [];
+		}
+
+		const scopes = await this.scopeRepository.findByList(scopeSlugs);
+		if (scopes.length !== scopeSlugs.length) {
+			const invalidScopes = scopeSlugs.filter((slug) => !scopes.some((s) => s.slug === slug));
+			throw new Error(`The following scopes are invalid: ${invalidScopes.join(', ')}`);
+		}
+
+		return scopes;
+	}
+
+	async updateCustomRole(slug: string, newData: UpdateRoleDto) {
+		const role = await this.roleRepository.findBySlug(slug);
+		if (!role) {
+			throw new NotFoundError('Role not found');
+		}
+		if (role.systemRole) {
+			throw new BadRequestError('Cannot update system roles');
+		}
+
+		const { displayName, description, scopes: scopeSlugs } = newData;
+
+		const updatedRole = await this.roleRepository.updateRole(slug, {
```

Comment:
we already have the role fetched from db here, and the updateRole method also tries to find it. I wonder if we could prevent double fetch